packages/node/src/indexer/unfinalizedBlocks.service.ts

```diff
@@ -0,0 +1,219 @@
+    return this.unfinalizedBlocks[number];
+  }
+
+  registerFinalizedBlock(block: SignedBlock): void {
+    if (
+      this.finalizedBlock &&
+      this.finalizedBlock.block.header.number.toNumber() ===
+        block.block.header.number.toNumber()
+    ) {
+      return;
+    }
+    this.finalizedBlock = block;
+  }
+
+  async registerUnfinalizedBlock(
+    blockNumber: number,
+    hash: HexString,
+    tx: Transaction,
+  ): Promise<void> {
+    if (
+      !this.unfinalizedBlocks[blockNumber] &&
+      blockNumber > this.finalizedBlock.block.header.number.toNumber()
+    ) {
+      this.storeUnfinalizedBlock(blockNumber, hash);
+    }
+    await this.saveUnfinalizedBlocks(this.unfinalizedBlocks, tx);
+  }
+
+  async deleteFinalizedBlock(tx: Transaction): Promise<void> {
+    if (
+      this.lastCheckedBlockHeight !== undefined &&
+      this.lastCheckedBlockHeight <
+        this.finalizedBlock.block.header.number.toNumber()
+    ) {
+      this.removeFinalized(this.finalizedBlock.block.header.number.toNumber());
+      await this.saveLastFinalizedVerifiedHeight(
+        this.finalizedBlock.block.header.number.toNumber(),
+        tx,
+      );
+      await this.saveUnfinalizedBlocks(this.unfinalizedBlocks, tx);
+    }
+    this.lastCheckedBlockHeight =
+      this.finalizedBlock.block.header.number.toNumber();
+  }
+
+  storeUnfinalizedBlock(blockNumber: number, hash: HexString): void {
+    this.unfinalizedBlocks[blockNumber] = hash;
+  }
+
+  private async saveUnfinalizedBlocks(
+    unfinalizedBlocks: Record<number, HexString>,
+    tx: Transaction,
+  ): Promise<void> {
+    return this.setMetadata(
+      METADATA_UNFINALIZED_BLOCKS_KEY,
+      JSON.stringify(unfinalizedBlocks),
+      tx,
+    );
+  }
+
+  private async saveLastFinalizedVerifiedHeight(
+    height: number,
+    tx: Transaction,
+  ) {
+    return this.setMetadata(METADATA_LAST_FINALIZED_PROCESSED_KEY, height, tx);
+  }
+
+  // remove any records less and equal than input finalized blockHeight
+  removeFinalized(blockHeight: number): void {
+    Object.entries(this.unfinalizedBlocks).map(([bestBlockHeight, hash]) => {
+      if (Number(bestBlockHeight) <= blockHeight) {
+        delete this.unfinalizedBlocks[bestBlockHeight];
+      }
+    });
+  }
+
+  // find closest record from block heights
+  getClosestRecord(
+    blockHeight: number,
+  ): { blockHeight: number; hash: HexString } | undefined {
+    // Have the block in the best block, can be verified
+    if (Object.keys(this.unfinalizedBlocks).length !== 0) {
+      const record = this.getSortedUnfinalizedBlocks().find(
+        ([bestBlockHeight, hash]) => Number(bestBlockHeight) <= blockHeight,
+      );
+      if (record) {
+        const [bestBlockHeight, hash] = record;
+        return { blockHeight: Number(bestBlockHeight), hash: hash };
+      }
+    }
+    return undefined;
+  }
+
+  // verify best blocks with finalized block
+  async validateUnfinalizedBlocks(): Promise<boolean> {
```

Comment:
Not sure if we need a seperate `validateUnfinalizedBlocks()`, can we do everything in one function instead? (find  fork, if yes, return fork height and purge unfinalized block array, if no, purge unfinalized block array to latest finalised height.